newrelic/resource_newrelic_api_access_key.go

```diff
@@ -0,0 +1,317 @@
+
+	// Define initial keys to create an API access key.
+	opts := apiaccesskeys.APIAccessCreateKeysInput{}
+
+	// Get the account id. This is required regardless of key type.
+	var accountId int
+	if v, ok := d.GetOk("account_id"); ok {
+		accountId = v.(int)
+		log.Printf("[DEBUG] new api access account_id: %d", accountId)
+	}
+
+	// Get the key type.
+	keyType := getAPIAccessKeyType(d)
+
+	// Validate to make sure the following:
+	// - If key_type is set to INGEST, `ingest_type` must be set to a value.
+	// - If key_type is set to USER, `user_id` must be set to a value.
+	switch keyType {
+	case KeyTypeIngest:
+		ingestKeyOpts := apiaccesskeys.APIAccessCreateIngestKeyInput{}
+
+		if v, ok := d.GetOk("ingest_type"); ok {
+			ingestKeyOpts.IngestType = v.(string)
+			log.Printf("[DEBUG] new api access ingest_type: %s", ingestKeyOpts.IngestType)
+		} else {
+			return errors.New("[ERROR] you must define the ingest_type attribute when creating an INGEST key")
+		}
+
+		ingestKeyOpts.AccountID = accountId
+		ingestKeyOpts.Name = getAPIAccessKeyName(d)
+		ingestKeyOpts.Notes = getAPIAccessKeyNotes(d)
+		opts.Ingest = []apiaccesskeys.APIAccessCreateIngestKeyInput{ingestKeyOpts}
+	case KeyTypeUser:
+		userKeyOpts := apiaccesskeys.APIAccessCreateUserKeyInput{}
+
+		if v, ok := d.GetOk("user_id"); ok {
+			userKeyOpts.UserID = v.(int)
+			log.Printf("[DEBUG] new api access user_id: %d", userKeyOpts.UserID)
+		} else {
+			return errors.New("[ERROR] you must define the user_id attribute when creating an USER key")
+		}
+
+		userKeyOpts.AccountID = accountId
+		userKeyOpts.Name = getAPIAccessKeyName(d)
+		userKeyOpts.Notes = getAPIAccessKeyNotes(d)
+		opts.User = []apiaccesskeys.APIAccessCreateUserKeyInput{userKeyOpts}
+	default:
+		return errors.New(fmt.Sprintf("unknown api access key type: %s", keyType))
+	}
+
+	keys, createErr := client.APIAccessKeys.CreateAPIAccessKeysMutation(opts)
+	if createErr != nil {
+		return createErr
+	}
+
+	// Validate to make sure we only created one key.
+	if len(keys) != 1 {
+		return errors.New(fmt.Sprintf("expected 1 new key, got %d", len(keys)))
+	}
+
+	// Set the resource ID to be a composite of the key ID and the key type in order to lookup the newly created key
+	d.SetId(keys[0].ID)
+
+	return resourceNewRelicAPIAccessKeyRead(d, meta)
+}
+
+func resourceNewRelicAPIAccessKeyRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ProviderConfig).NewClient
+
+	key, readErr := client.APIAccessKeys.GetAPIAccessKeyMutation(apiaccesskeys.APIAccessGetInput{
+		ID: d.Id(), KeyType: getAPIAccessKeyType(d),
+	})
+	if readErr != nil {
+		return readErr
+	}
+
+	var setErr error
+	setErr = d.Set("account_id", key.AccountID)
+	if setErr != nil {
+		return setErr
+	}
+
+	setErr = d.Set("key_type", key.Type)
+	if setErr != nil {
+		return setErr
+	}
+
+	setErr = d.Set("ingest_type", key.IngestType)
+	if setErr != nil {
+		return setErr
+	}
+
+	// TODO: add this back in
+	//setErr = d.Set("user_id", key.UserID)
```

Comment:
Did you still want to include this?

Reply:
Ah yes, once Sanders merges his PR in the `newrelic-client-go`, I'll add this line back.

I commented it out to test the INGEST keys.

Reply:
Wow, you're making moves fast @davidji99. I just wrapped up the [newrelic-client-go fast-follow pull request](https://github.com/newrelic/newrelic-client-go/pull/433) I mentioned that would be on the way. That PR contains several updates to the types and method signatures in the client. My apologies in advance as this will require some refactoring in this PR. Hopefully it's nothing too crazy. And please feel free to review and provide feedback on the client PR as well 😎

Reply:
Yeah! Our team really needs this for programmatic key rotations. Also kudos to New Relic for adding API support for keys. Let me pull in your branch and do some refactoring to test out https://github.com/newrelic/newrelic-client-go/pull/433.